packages/x-date-pickers/src/CalendarPicker/DayPicker.tsx

```diff
@@ -365,6 +416,16 @@ export function DayPicker<TDate>(inProps: DayPickerProps<TDate>) {
   return (
     <div role="grid" aria-labelledby={gridLabelId}>
       <PickersCalendarDayHeader role="row" className={classes.header}>
+        {displayWeekNumber && getWeekNumber && (
+          <PickersCalendarWeekNumberLabel
+            variant="caption"
+            role="columnheader"
+            aria-label={localeText.calendarWeekNumberHeaderLabel}
```

Comment:
Have you tested a11y behaviour with any tool?
I've tried this implementation with NVDA on Windows and it does not seem to pick up this relationship between this `columnheader` and the `rowheader` on L469.
Also, with `week number` the screen reader assumes `8` columns, whereas it's actually still 7 with a `describing/header` week column.

FYI. I'm indifferent on what to do with `aria-label` on L423 as of right now it's value does not seem to be used, but in theory it makes sense to have it... 🤷 🤔 
Maybe @m4theushw has encountered such a11y case of having a non-interactive describing grid/table column and knows how to best implement it?

I'm proposing the following solution to allow for correct `week` and column count announcement when navigating the days.

```diff
diff --git a/packages/x-date-pickers/src/CalendarPicker/DayPicker.tsx b/packages/x-date-pickers/src/CalendarPicker/DayPicker.tsx
index 961bebaa..6e241d92 100644
--- a/packages/x-date-pickers/src/CalendarPicker/DayPicker.tsx
+++ b/packages/x-date-pickers/src/CalendarPicker/DayPicker.tsx
@@ -459,18 +459,26 @@ export function DayPicker<TDate>(inProps: DayPickerProps<TDate>) {
             role="rowgroup"
             className={classes.monthContainer}
           >
-            {utils.getWeekArray(currentMonth).map((week) => (
+            {utils.getWeekArray(currentMonth).map((week, index) => (
               <PickersCalendarWeek
                 role="row"
                 key={`week-${week[0]}`}
+                // fix existing issue of announcing first row as row 2
+                aria-rowindex={index + 1}
                 className={classes.weekContainer}
               >
                 {displayWeekNumber && getWeekNumber && (
-                  <PickersCalendarWeekNumber className={classes.weekNumber} role="rowheader">
+                  <PickersCalendarWeekNumber
+                    className={classes.weekNumber}
+                    role="rowheader"
+                    aria-label={localeText.calendarWeekNumberColumnHeaderText(
+                      getWeekNumber(week[0]),
+                    )}
+                  >
                     {getWeekNumber(week[0])}
                   </PickersCalendarWeekNumber>
                 )}
-                {week.map((day) => {
+                {week.map((day, index) => {
                   const isFocusableDay =
                     focusableDay !== null && utils.isSameDay(day, focusableDay);
                   const isSelected = validSelectedDays.some((selectedDay) =>
@@ -495,6 +503,7 @@ export function DayPicker<TDate>(inProps: DayPickerProps<TDate>) {
                     tabIndex: isFocusableDay ? 0 : -1,
                     role: 'gridcell',
                     'aria-selected': isSelected,
+                    'aria-colindex': index + 1,
                   };
                   if (isToday) {
                     pickersDayProps['aria-current'] = 'date';
diff --git a/packages/x-date-pickers/src/locales/enUS.ts b/packages/x-date-pickers/src/locales/enUS.ts
index 23e58351..948fccf7 100644
--- a/packages/x-date-pickers/src/locales/enUS.ts
+++ b/packages/x-date-pickers/src/locales/enUS.ts
@@ -49,6 +49,7 @@ const enUSPickers: PickersLocaleText<any> = {
   // Calendar labels
   calendarWeekNumberHeaderLabel: 'Week number',
   calendarWeekNumberHeaderText: '#',
+  calendarWeekNumberColumnHeaderText: (weekNumber) => `Week ${weekNumber}`,
 
   // Open picker labels
   openDatePickerDialogue: (value, utils) =>
diff --git a/packages/x-date-pickers/src/locales/utils/pickersLocaleTextApi.ts b/packages/x-date-pickers/src/locales/utils/pickersLocaleTextApi.ts
index 2dc29945..0b988a2b 100644
--- a/packages/x-date-pickers/src/locales/utils/pickersLocaleTextApi.ts
+++ b/packages/x-date-pickers/src/locales/utils/pickersLocaleTextApi.ts
@@ -15,6 +15,7 @@ export interface PickersLocaleText<TDate> {
   end: string;
   calendarWeekNumberHeaderLabel: string;
   calendarWeekNumberHeaderText: string;
+  calendarWeekNumberColumnHeaderText: (weekNumber: string) => string;
   calendarViewSwitchingButtonAriaLabel: (currentView: CalendarPickerView) => string;
   inputModeToggleButtonAriaLabel: (
     isKeyboardInputOpen: boolean,
```

Reply:
I tried to respect sementic meaning but did not tested.

I searched a bit about how other libraries do. A solution is to use `aria-hidden` on the header and the week number to have the correct number of row/columns. But then you have to adapt the `aria-label` of each button such that they provide the day

Reply:
Yeah, I also thought/tried `aria-hidden` approach to the non-interactive row/column, but it seemed like a really not cool hack. 🙈 
IMO, `aria-colindex` and `aria-rowindex` are way less intrusive way to achieve what we want.
FYI, possibly using a `table` would be more accessible by screen readers, haven't looked into it, but even MDN specifies that if possible, table layout should be preferred instead of replicating same behaviour with necessary `role` elements.

Reply:
I agree but not sure it's good to do a feature and an big accessibility update in the same PR

Reply:
Done 👍